### PR TITLE
Upgrade to the latest operator-sdk

### PR DESCRIPTION
**Pull request by mprahl**
The first commit removed some unused code regarding namespace selectors. This made fixing the tests after the operator-sdk upgrade easier.

operator-sdk v1.x now wraps Kubebuilder for its scaffolding and code
generation. This changes the path of `./cmd/manager/main.go` to
`.main.go`, `./pkg/controller` to `./controllers`, and `pkg/apis` to
`api`.

Manifest generation is also handled by Kubebuilder. This is handled by
the `generate-operator-yaml` Make target. This means that any RBAC
changes must be done using kubebuilder markers and not manual edits to
the manifests in `./deploy`. The RBAC Kubebuilder markers have been
refined to follow the least privilege principal.

The largest changes stem from updating the underlying Kubernetes Go
modules as some APIs changed or were removed. This also leads to a new
way of performing leader election via leases. For old Kubernetes
versions such as OCP 3.11, the lease API is not available so a flag
called `--legacy-leader-election` will go back to the old way of
performing leader election.

It's also important to note that operator-sdk is no longer a Go module
dependency, so some functionality was copied over, mostly unmodified.

Lastly, the Policy scheme is now imported from the propagator rather
than redefined in this controller.

Relates:
open-cluster-management/backlog#15552

### CODE REVIEW

**Comment by mprahl** [comment]
The Dockerfile looks like a new file but it's just because I removed the symlink from Dockerfile to build/Dockerfile. This was a previous workaround for operator-sdk v0.19.4.

**Review by JustinKuli — CHANGES_REQUESTED** [review]
Looks good overall! Just one question and a small requested change

**Comment by mprahl** [comment]
@ckandag could I please get your input on the ACL refactoring? It works as expected based on my testing and running the framework tests on it but I'd like your opinion as well.

File: Makefile [hunk 1]
```diff
@@ -84,15 +84,15 @@ dependencies-go:
 	go mod download
 
 build:
-	CGO_ENABLED=0 GOOS=$(GOOS) GOARCH=$(GOARCH) go build -a -tags netgo -o ./build/_output/bin/iam-policy-controller ./cmd/manager
+	CGO_ENABLED=0 GOOS=$(GOOS) GOARCH=$(GOARCH) go build -a -tags netgo -o ./build/_output/bin/iam-policy-controller ./
 
 build-images:
-	@docker build -t ${IMAGE_NAME_AND_VERSION} -f ./build/Dockerfile .
+	@docker build -t ${IMAGE_NAME_AND_VERSION} .
```
[the remark this answers is not in the document]
Comment (JustinKuli) [reply Makefile:90]: Ok, now I see the prow config here: https://github.com/openshift/release/blob/master/ci-operator/config/open-cluster-management/iam-policy-controller/open-cluster-management-iam-policy-controller-release-2.4.yaml#L13, and that's pointing to `./Dockerfile`.

AFAIK, the other controllers still use `./build/Dockerfile`, for example: https://github.com/openshift/release/blob/master/ci-operator/config/open-cluster-management/config-policy-controller/open-cluster-management-config-policy-controller-release-2.4.yaml#L13

I guess it doesn't matter.

**Review by JustinKuli — APPROVED** [review]

**Review by gparvin — APPROVED** [review]
I'm not sure I understand how it needs only `clusterrolebindings` and not also `clusterroles`, but definitely glad to see least privilege is being improved.  Thanks for putting in the effort to improve that.

**Comment by mprahl** [comment]
> I'm not sure I understand how it needs only `clusterrolebindings` and not also `clusterroles`, but definitely glad to see least privilege is being improved. Thanks for putting in the effort to improve that.

I think we could tighten this up with a role (not cluster role) for how it accesses the hub config secret it needs for creating the lease since that should always be in the `open-cluster-management-agent-addon` namespace.

Since the Helm chart in the klusterlet-addon-operator just uses a cluster role, I'll keep it as is in this PR and file a separate issue for that.

**Pull request merged**